### PR TITLE
Add spec for IfOp

### LINKED ISSUE #311: Add spec for IfOp
### Request description

To spec stablehlo.if.

### PR DESCRIPTION
**Pull request by subhankarshah**
closes #311

### CODE REVIEW

File: docs/spec_draft.md [hunk 1]
```diff
@@ -680,6 +681,48 @@ IEEE-754 specification.
...
+
+### Outputs
+
+| Name     | Type               |
+|----------|--------------------|
+| `result` | any supported type |
+
+### Constraints
+
+  * (C1) `true_branch` and `false_branch` take 0 arguments.
```
Comment (sdasgup3) [inline docs/spec_draft.md:707]: At the top we are defining what a function is. IMO, we should say that the function can capture implicit arguments from the parent scope.

Comment (burmako) [reply docs/spec_draft.md:707]: Later today, I was planning to send a PR that describes function semantics. Will ping it here when done.

Comment (subhankarshah) [reply docs/spec_draft.md:707]: @sdasgup3 yes, I was thinking about that as well, as in the functions have 0 arguments but they can use whatever is in the scope of parent function.

File: docs/spec_draft.md [hunk 2]
```diff
@@ -680,6 +681,48 @@ IEEE-754 specification.
...
+|----------------|--------------------------------------------|
+| `pred`         | 1-dimensional tensor constant of type `i1` |
+| `true_branch`  | function of type `function`                |
+| `false_branch` | function of type `function`                |
+
+### Outputs
+
+| Name     | Type               |
+|----------|--------------------|
+| `result` | any supported type |
```
[the remark this answers is not in the document]
Comment (subhankarshah) [reply docs/spec_draft.md:703]: they can be anything, not necessarily a variadic number of tensors

[the remark this answers is not in the document]
Comment (burmako) [reply docs/spec_draft.md:703]: "variadic number of tensors of any supported type" sounds like the right wording to me as well. We don't have ops which return individual element types, and "variadic" is not a type in our type system, so "variadic number of tensors" is as general as we can go.

File: docs/spec_draft.md [hunk 3]
```diff
@@ -680,6 +681,48 @@ IEEE-754 specification.
...
+Returns the result of `true_branch` if `pred` is `true`, `false_branch` if
+`pred` is `false`.
+
+### Inputs
+
+| Name           | Type                                       |
+|----------------|--------------------------------------------|
+| `pred`         | 1-dimensional tensor constant of type `i1` |
+| `true_branch`  | function of type `function`                |
+| `false_branch` | function of type `function`                |
```
Comment (burmako) [inline docs/spec_draft.md:697]: Since this column says "Type", I think this should be just "\`function\`".

File: docs/spec_draft.md [hunk 4]
```diff
@@ -680,6 +681,48 @@ IEEE-754 specification.
...
+### Outputs
+
+| Name     | Type               |
+|----------|--------------------|
+| `result` | any supported type |
+
+### Constraints
+
+  * (C1) `true_branch` and `false_branch` take 0 arguments.
+  * (C2) Return types of `true_branch` and `false_branch` are same.
```
Comment (burmako) [inline docs/spec_draft.md:708]: Let's reword the constraints based on the definition of "function type" from https://github.com/openxla/stablehlo/blob/main/docs/spec_draft.md#types and of "function" from https://github.com/openxla/stablehlo/blob/main/docs/spec_draft.md#programs, i.e.

```
  * (C1) `true_branch` and `false_branch` have 0 inputs.
  * (C2) `true_branch` and `false_branch` have the same output types.
```

File: docs/spec_draft.md [hunk 5]
```diff
@@ -680,6 +681,48 @@ IEEE-754 specification.
 
 [Back to Ops](#index-of-ops)
 
+## stablehlo.if
+
+### Semantics
+
+Returns the result of `true_branch` if `pred` is `true`, `false_branch` if
+`pred` is `false`.
```
Comment (burmako) [inline docs/spec_draft.md:689]: How about "Produces the output of"?
  * "Produces" follows the convention that other ops use and avoids ambiguity with `return`.
  * "the output of" matches how functions are defined in terms of inputs and outputs in the spec.

Comment (burmako) [inline docs/spec_draft.md:689]: Also, let's call out that only one branch is executed - this is material to the semantics of the op. E.g. "Only one branch is executed depending on the value of `pred`".

File: stablehlo/dialect/StablehloOps.td [hunk 6]
```diff
@@ -1142,15 +1142,18 @@ def StableHLO_IfOp: StableHLO_Op<"if", [
   let summary = "If operator";
```
Comment (burmako) [inline stablehlo/dialect/StablehloOps.td:1142]: Let's also delete the comment above `def StableHLO_IfOp`. It is not relevant to StableHLO.

File: docs/spec_draft.md (hunk 4 above)
Comment (burmako) [inline docs/spec_draft.md:708]: This also needs "(C3) For all `i`, `type(results[i])` = `type(true_branch).outputs[i]`.

**Review by burmako — CHANGES_REQUESTED** [review]

File: docs/spec_draft.md (hunk 5 above)
Comment (subhankarshah) [reply docs/spec_draft.md]: Produces the output from executing exactly one branch from `true_branch` or
`false_branch` depending on the value of `pred`. Formally, if `pred` is `true`,
output of `true_branch` is returned, else if pred is `false`, output of
`false_branch` is returned.

Split it into two lines, still using the word `return` in second line.

Comment (subhankarshah) [reply docs/spec_draft.md]: Captured this in first line of two lines.

File: docs/spec_draft.md [hunk 7]
```diff
@@ -691,6 +692,51 @@ IEEE-754 specification.
...
+
+| Name     | Type                                             |
+|----------|--------------------------------------------------|
+| `result` | variadic number of tensors of any supported type |
+
+### Constraints
+
+  * (C1) `true_branch` and `false_branch` have 0 inputs.
+  * (C2) `true_branch` and `false_branch` have the same output types.
+  * (C3) For all `i`, `type(result[i]) = type(branches[0]).outputs[i]`.
```
Comment (burmako) [inline docs/spec_draft.md:722]: "type(true_branch)"

**Review by burmako — CHANGES_REQUESTED** [review]
One minor change, and we're good to go.

File: docs/spec_draft.md [hunk 8]
```diff
@@ -691,6 +692,51 @@ IEEE-754 specification.
...
+|----------------|--------------------------------------------|
+| `pred`         | 1-dimensional tensor constant of type `i1` |
+| `true_branch`  | `function`                                 |
+| `false_branch` | `function`                                 |
+
+### Outputs
+
+| Name     | Type                                             |
+|----------|--------------------------------------------------|
+| `result` | variadic number of tensors of any supported type |
```
Comment (burmako) [inline docs/spec_draft.md:716]: As discussed in the CaseOp PR, this should be called "results". Nice catch!

**Review by burmako — CHANGES_REQUESTED** [review]

**Review by burmako — APPROVED** [review]

**Review by ghpvnist — APPROVED** [review]

File: docs/spec_draft.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (burmako) [reply docs/spec_draft.md]: Looks like this comment has already been addressed? The spec currently says "(C2) `true_branch` and `false_branch` have the same output types".

**Pull request merged**